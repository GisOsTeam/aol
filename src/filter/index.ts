/**
 * @file Automatically generated by barrelsby.
 */

export * from './FilterBuilder';
export * from './IFilter';
