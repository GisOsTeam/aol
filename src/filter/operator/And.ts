import { IOperator, OperatorEnum, OperatorType } from './IOperator';

export class And implements IOperator {
  public not: boolean;
  public readonly type: OperatorType = OperatorEnum.and;

  constructor(not = false) {
    this.not = not;
  }

  public toString(): string {
    if (this.not) {
      return '&& !';
    }
    return '&&';
  }
}
