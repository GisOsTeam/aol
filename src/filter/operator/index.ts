/**
 * @file Automatically generated by barrelsby.
 */

export * from './AndOp';
export * from './Equal';
export * from './GreaterOrEqualThan';
export * from './GreaterThan';
export * from './IOperator';
export * from './Ilike';
export * from './In';
export * from './Like';
export * from './LowerOrEqualThan';
export * from './LowerThan';
export * from './Null';
export * from './Or';
