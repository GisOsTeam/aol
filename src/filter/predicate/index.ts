/**
 * @file Automatically generated by barrelsby.
 */

export * from './AndPre';
export * from './BasePredicate';
export * from './Equal';
export * from './FilterPredicate';
export * from './GreaterOrEqualThan';
export * from './GreaterThan';
export * from './IPredicate';
export * from './Ilike';
export * from './In';
export * from './Like';
export * from './LowerOrEqualThan';
export * from './LowerThan';
export * from './Or';
