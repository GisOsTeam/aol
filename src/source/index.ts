/**
 * @file Automatically generated by barrelsby.
 */

export * from './ExternalVector';
export * from './Image';
export * from './ImageArcGISRest';
export * from './ImageStatic';
export * from './ImageWms';
export * from './LocalVector';
export * from './QueryArcGISRest';
export * from './TileArcGISRest';
export * from './TileImage';
export * from './TileWms';
export * from './Vector';
export * from './VectorTile';
export * from './Wfs';
export * from './Xyz';
export * from './query/index';
export * from './types/index';
