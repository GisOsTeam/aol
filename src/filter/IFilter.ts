import { IOperator } from './operator';

export enum FieldTypeEnum {
  Oid = 'Oid',
  Boolean = 'Boolean',
  Number = 'Number',
  String = 'String',
  Date = 'Date',
  Geometry = 'Geometry',
  Unknown = 'Unknown',
}

export type FieldType =
  | FieldTypeEnum.Oid
  | FieldTypeEnum.Boolean
  | FieldTypeEnum.Number
  | FieldTypeEnum.String
  | FieldTypeEnum.Date
  | FieldTypeEnum.Geometry
  | FieldTypeEnum.Unknown;

export interface IField<T> {
  key: string | keyof T;
  type: FieldType;
  alias?: string;
}

export interface IFilter<T> {
  field: IField<T>;
  operator: IOperator;
  value: string | number | boolean | any[];
}

export interface IPredicate<T> extends IFilter<T> {
  toString(filterBuilderType?: FilterBuilderType): string;
}

export enum FilterBuilderTypeEnum {
  CQL = 'CQL',
  SQL = 'SQL',
  OGC = 'OGC',
}

export type FilterBuilderType = FilterBuilderTypeEnum.CQL | FilterBuilderTypeEnum.SQL | FilterBuilderTypeEnum.OGC;
