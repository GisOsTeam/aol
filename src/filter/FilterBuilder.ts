import { FilterBuilderType } from './IFilter';
import { AndPre, IPredicate, Or } from './predicate';

export class FilterBuilder {
  public predicate: IPredicate;

  constructor(predicate: IPredicate) {
    this.predicate = predicate;
  }

  public static build(predicate: IPredicate, type: FilterBuilderType): string {
    if (!predicate) {
      return 'INCLUDE';
    }
    return predicate.toString(type);
  }

  public build(type: FilterBuilderType): string {
    if (!this.predicate) {
      return 'INCLUDE';
    }
    return this.predicate.toString(type);
  }

  public and<RP extends IPredicate>(rightPredicate: RP): FilterBuilder {
    return new FilterBuilder(new AndPre(this.predicate, rightPredicate));
  }

  public or<RP extends IPredicate>(rightPredicate: RP): FilterBuilder {
    return new FilterBuilder(new Or<any, RP>(this.predicate, rightPredicate));
  }
}
