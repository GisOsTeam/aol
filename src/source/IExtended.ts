import OlMap from 'ol/Map';
import Source, { Options } from 'ol/source/Source';
import Feature from 'ol/Feature';
import Geometry from 'ol/geom/Geometry';
import Projection from 'ol/proj/Projection';
import { SourceType } from './types/sourceType';
import { LayerType } from './types/layerType';
import { AndPre, IPredicate, Or } from '../filter/predicate';
import { FilterBuilder } from '../filter/FilterBuilder';

export interface ISnapshotOptions extends Options {
  snapshotable?: boolean;
  listable?: boolean;
  removable?: boolean;
}

export interface ISnapshotSource extends Source {
  getSourceType(): SourceType;

  getSourceOptions(): ISnapshotOptions;

  setSourceOptions(options: ISnapshotOptions): void;

  getLayerType(): LayerType;

  isSnapshotable(): boolean; // Can save snapshot
  isListable(): boolean; // Visible by others tools
  isRemovable(): boolean; // Removable from map by others tools
}

export interface IInitSource extends ISnapshotSource {
  init(): Promise<void>;
}

export interface IQuerySource extends ISnapshotSource {
  query(identifyRequest: IGisRequest, onlyVisible?: boolean): Promise<IQueryResponse>;

  retrieveFeature(id: number | string, featureProjection: Projection): Promise<Feature>;
}

export interface ILayerLegend {
  /**
   * Label optinal
   */
  label?: string;
  /**
   * Image element
   */
  image: HTMLImageElement;
  /**
   * Image data in Base64 ou url
   */
  srcImage: string;
  /**
   * Height
   */
  height: number;
  /**
   * Width
   */
  width: number;
}

export type ILegendRecord = Record<number | string, ILayerLegend[]>;

export interface ILegendSource {
  fetchLegend(): Promise<ILegendRecord>;
}

export interface IExtended extends IInitSource, IQuerySource, ILegendSource { }

export type IGisRequest = IQueryRequest | IIdentifyRequest;

export interface IAbstractRequest<T extends string> {
  olMap: OlMap;
  queryType: T;
  filters?: IPredicate;
  limit?: number;
  returnGeometry?: boolean;
  srId?: string;
}

export interface IQueryRequest extends IAbstractRequest<'query'> {
  geometry?: Geometry;
  geometryProjection?: Projection;
  geometryPrecision?: number;
}

export interface IIdentifyRequest extends IAbstractRequest<'identify'> {
  geometry: Geometry;
  geometryPrecision?: number;
  geometryProjection: Projection;
  identifyTolerance?: number; // in pixel
  layersPrefix?: LayersPrefix;
  outFields?: string;
  returnFieldName?: boolean;
}

export interface IQueryResponse {
  request: IGisRequest;
  featureTypeResponses: IQueryFeatureTypeResponse[];
}

export interface IQueryFeatureTypeResponse {
  type: IFeatureType<any>;
  features: Feature[];
  source: IQuerySource;
}

export interface IAttribute {
  key: string;
  type?: 'Oid' | 'Boolean' | 'Number' | 'String' | 'Date' | 'Geometry' | 'Unknown';
  name?: string;
}

export interface IFeatureType<IDT extends number | string> {
  id: IDT;
  queryId?: IDT;
  hide?: boolean;
  name?: string;
  identifierAttribute?: IAttribute;
  attributes?: IAttribute[];
  readonly initialPredicate?: IPredicate;

  joinWithOr?: boolean;

  predicateWrapper?: FeatureTypePredicateWrapper;
}

export class FeatureTypePredicateWrapper {
  public joinWithOr: boolean;

  private _initialPredicate: IPredicate;
  private _additionalPredicates: Record<string, IPredicate> = {};

  public constructor(initialPredicate?: IPredicate, joinWithOr = true) {
    this._initialPredicate = initialPredicate;
    this.joinWithOr = joinWithOr;
  }

  public addPredicate(predicate: IPredicate): void {
    this._additionalPredicates[predicate.hashCode()] = predicate;
  }

  public removePredicate(hash: string) {
    delete this._additionalPredicates[hash];
  }

  get predicate(): IPredicate {
    return this.filterBuilder.predicate;
  }

  get filterBuilder(): FilterBuilder {
    const predicates: IPredicate[] = Object.values(this._additionalPredicates);
    if (predicates.length > 0) {
      let filterBuilder: FilterBuilder;
      for (const predicate of predicates) {
        if(!filterBuilder) {
          filterBuilder = new FilterBuilder(predicate);
        } else {
          if (this.joinWithOr) {
            filterBuilder.or(predicate);
          } else {
            filterBuilder.and(predicate);
          }
        }
      }
      return filterBuilder.and(this._initialPredicate);
    }
    return new FilterBuilder(this._initialPredicate);
  }

}
export type LayersPrefix = LayersPrefixEnum.ALL | LayersPrefixEnum.TOP | LayersPrefixEnum.VISIBLE;
export enum LayersPrefixEnum {
  ALL = 'all',
  TOP = 'top',
  VISIBLE = 'visible',
}
