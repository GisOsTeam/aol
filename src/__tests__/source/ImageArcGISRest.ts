import { ImageArcGISRest } from '../../source';
import { Equal, IPredicate, Like } from '../../filter/predicate';
import { Equal as EqualOp, Like as LikeOp } from '../../filter/operator';
import { FieldTypeEnum, FilterBuilder, IField } from '../../filter';
import Projection from 'ol/proj/Projection';
import ImageWrapper from 'ol/Image';
import { IFeatureType } from '../../source/IExtended';

describe('aol.source.imageArcGISRest', () => {
  describe('ArcGIS Online', () => {
    const url = 'https://sampleserver6.arcgisonline.com/arcgis/rest/services/USA/MapServer';
    const sourceOptions = {
      url,
      types: [
        { id: 1, identifierAttribute: { key: 'objectid' } },
        { id: 2, identifierAttribute: { key: 'objectid' } },
      ],
      ratio: 1,
      params: { TRANSPARENT: false },
    };

    const imageLoadFunction = (image: ImageWrapper, src: string) => {
      expect(src).toMatchSnapshot();
    };

    const field: IField<any> = { type: FieldTypeEnum.String, key: 'state_name' };
    const equal: IPredicate = new Equal(field, new EqualOp(), 'Washington');
    const like: IPredicate = new Like(field, new LikeOp(), '%Dako%');
    const or: IPredicate = new FilterBuilder(equal).or(like).predicate;

    describe('filter', () => {
      describe('1 - constructor', () => {
        test('equal', async () => {
          const types: IFeatureType<number>[] = [
            { id: 1, identifierAttribute: { key: 'objectid' } },
            { id: 2, identifierAttribute: { key: 'objectid' }, predicate: equal },
          ];
          const imageArcGISRest = new ImageArcGISRest({ ...sourceOptions, ...{ types } });
          imageArcGISRest.setImageLoadFunction(imageLoadFunction);
          loadImageFromImageArcGISRest(imageArcGISRest);
        });
        test('like', async () => {
          const types: IFeatureType<number>[] = [
            { id: 1, identifierAttribute: { key: 'objectid' } },
            { id: 2, identifierAttribute: { key: 'objectid' }, predicate: like },
          ];
          const imageArcGISRest = new ImageArcGISRest({ ...sourceOptions, ...{ types } });
          imageArcGISRest.setImageLoadFunction(imageLoadFunction);
          loadImageFromImageArcGISRest(imageArcGISRest);
        });
        test('or', async () => {
          const types: IFeatureType<number>[] = [
            { id: 1, identifierAttribute: { key: 'objectid' } },
            { id: 2, identifierAttribute: { key: 'objectid' }, predicate: or },
          ];
          const imageArcGISRest = new ImageArcGISRest({ ...sourceOptions, ...{ types } });
          imageArcGISRest.setImageLoadFunction(imageLoadFunction);
          loadImageFromImageArcGISRest(imageArcGISRest);
        });
      });
      describe('2 - setSourceOptions', () => {
        const imageArcGISRest = new ImageArcGISRest({ ...sourceOptions });
        imageArcGISRest.setImageLoadFunction(imageLoadFunction);
        test('equal', async () => {
          const types: IFeatureType<number>[] = [
            { id: 1, identifierAttribute: { key: 'objectid' } },
            { id: 2, identifierAttribute: { key: 'objectid' }, predicate: equal },
          ];
          imageArcGISRest.setSourceOptions({ ...sourceOptions, ...{ types } });
          loadImageFromImageArcGISRest(imageArcGISRest);
        });
        test('like', async () => {
          const types: IFeatureType<number>[] = [
            { id: 1, identifierAttribute: { key: 'objectid' } },
            { id: 2, identifierAttribute: { key: 'objectid' }, predicate: like },
          ];
          imageArcGISRest.setSourceOptions({ ...sourceOptions, ...{ types } });
          loadImageFromImageArcGISRest(imageArcGISRest);
        });
        test('or', async () => {
          const types: IFeatureType<number>[] = [
            { id: 1, identifierAttribute: { key: 'objectid' } },
            { id: 2, identifierAttribute: { key: 'objectid' }, predicate: or },
          ];
          imageArcGISRest.setSourceOptions({ ...sourceOptions, ...{ types } });
          loadImageFromImageArcGISRest(imageArcGISRest);
        });
      });
      describe('3 - both', () => {
        const types: IFeatureType<number>[] = [
          { id: 1, identifierAttribute: { key: 'objectid' } },
          { id: 2, identifierAttribute: { key: 'objectid' }, predicate: like },
        ];
        const imageArcGISRest = new ImageArcGISRest({ ...sourceOptions, ...{ types } });
        imageArcGISRest.setImageLoadFunction(imageLoadFunction);

        test('', async () => {
          const like2: IPredicate = new Like(field, new LikeOp(), 'North%');
          const types: IFeatureType<number>[] = [{ id: 2, identifierAttribute: { key: 'objectid' }, predicate: like2 }];
          imageArcGISRest.setSourceOptions({ ...sourceOptions, ...{ types } });
          loadImageFromImageArcGISRest(imageArcGISRest);
        });

        test('', async () => {
          const types: IFeatureType<number>[] = [{ id: 2, identifierAttribute: { key: 'objectid' }, predicate: equal }];
          imageArcGISRest.setSourceOptions({ ...sourceOptions, ...{ types } });
          loadImageFromImageArcGISRest(imageArcGISRest);
        });
      });
    });
  });
});

const loadImageFromImageArcGISRest = (imageArcGISRest: ImageArcGISRest): void => {
  const projection = new Projection({ code: 'EPSG:4326' });
  imageArcGISRest
    .getImage([-173.24789851593732, -69.43544422133593, -16.2692137558099, 109.10900234289849], 1, 1, projection)
    .load();
};
