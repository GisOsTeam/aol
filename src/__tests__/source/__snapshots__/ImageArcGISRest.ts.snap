// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`aol.source.imageArcGISRest ArcGIS Online filter 1 - constructor equal 1`] = `"https://sampleserver6.arcgisonline.com/arcgis/rest/services/USA/MapServer/export?F=image&FORMAT=PNG32&TRANSPARENT=false&LAYERS=show%3A1%2C2&now=1487076708000&LAYERDEFS=%7B%222%22%3A%22(state_name%20%3D%20'Washington')%22%7D&SIZE=157%2C179&BBOX=-173.25855613587362%2C-69.66322093921872%2C-16.25855613587362%2C109.33677906078128&BBOXSR=4326&IMAGESR=4326&DPI=90"`;

exports[`aol.source.imageArcGISRest ArcGIS Online filter 1 - constructor like 1`] = `"https://sampleserver6.arcgisonline.com/arcgis/rest/services/USA/MapServer/export?F=image&FORMAT=PNG32&TRANSPARENT=false&LAYERS=show%3A1%2C2&now=1487076708000&LAYERDEFS=%7B%222%22%3A%22(state_name%20LIKE%20'%25Dako%25')%22%7D&SIZE=157%2C179&BBOX=-173.25855613587362%2C-69.66322093921872%2C-16.25855613587362%2C109.33677906078128&BBOXSR=4326&IMAGESR=4326&DPI=90"`;

exports[`aol.source.imageArcGISRest ArcGIS Online filter 1 - constructor or 1`] = `"https://sampleserver6.arcgisonline.com/arcgis/rest/services/USA/MapServer/export?F=image&FORMAT=PNG32&TRANSPARENT=false&LAYERS=show%3A1%2C2&now=1487076708000&LAYERDEFS=%7B%222%22%3A%22((state_name%20%3D%20'Washington')%20OR%20(state_name%20LIKE%20'%25Dako%25'))%22%7D&SIZE=157%2C179&BBOX=-173.25855613587362%2C-69.66322093921872%2C-16.25855613587362%2C109.33677906078128&BBOXSR=4326&IMAGESR=4326&DPI=90"`;

exports[`aol.source.imageArcGISRest ArcGIS Online filter 2 - setSourceOptions equal 1`] = `"https://sampleserver6.arcgisonline.com/arcgis/rest/services/USA/MapServer/export?F=image&FORMAT=PNG32&TRANSPARENT=false&LAYERS=show%3A1%2C2&now=1487076708000&LAYERDEFS=%7B%222%22%3A%22(state_name%20%3D%20'Washington')%22%7D&SIZE=157%2C179&BBOX=-173.25855613587362%2C-69.66322093921872%2C-16.25855613587362%2C109.33677906078128&BBOXSR=4326&IMAGESR=4326&DPI=90"`;

exports[`aol.source.imageArcGISRest ArcGIS Online filter 2 - setSourceOptions like 1`] = `"https://sampleserver6.arcgisonline.com/arcgis/rest/services/USA/MapServer/export?F=image&FORMAT=PNG32&TRANSPARENT=false&LAYERS=show%3A1%2C2&now=1487076708000&LAYERDEFS=%7B%222%22%3A%22(state_name%20LIKE%20'%25Dako%25')%22%7D&SIZE=157%2C179&BBOX=-173.25855613587362%2C-69.66322093921872%2C-16.25855613587362%2C109.33677906078128&BBOXSR=4326&IMAGESR=4326&DPI=90"`;

exports[`aol.source.imageArcGISRest ArcGIS Online filter 2 - setSourceOptions or 1`] = `"https://sampleserver6.arcgisonline.com/arcgis/rest/services/USA/MapServer/export?F=image&FORMAT=PNG32&TRANSPARENT=false&LAYERS=show%3A1%2C2&now=1487076708000&LAYERDEFS=%7B%222%22%3A%22((state_name%20%3D%20'Washington')%20OR%20(state_name%20LIKE%20'%25Dako%25'))%22%7D&SIZE=157%2C179&BBOX=-173.25855613587362%2C-69.66322093921872%2C-16.25855613587362%2C109.33677906078128&BBOXSR=4326&IMAGESR=4326&DPI=90"`;

exports[`aol.source.imageArcGISRest ArcGIS Online filter 3 - both  1`] = `"https://sampleserver6.arcgisonline.com/arcgis/rest/services/USA/MapServer/export?F=image&FORMAT=PNG32&TRANSPARENT=false&LAYERS=show%3A2&now=1487076708000&LAYERDEFS=%7B%222%22%3A%22((state_name%20LIKE%20'%25Dako%25')%20AND%20(state_name%20LIKE%20'North%25'))%22%7D&SIZE=157%2C179&BBOX=-173.25855613587362%2C-69.66322093921872%2C-16.25855613587362%2C109.33677906078128&BBOXSR=4326&IMAGESR=4326&DPI=90"`;

exports[`aol.source.imageArcGISRest ArcGIS Online filter 3 - both  2`] = `"https://sampleserver6.arcgisonline.com/arcgis/rest/services/USA/MapServer/export?F=image&FORMAT=PNG32&TRANSPARENT=false&LAYERS=show%3A2&now=1487076708000&LAYERDEFS=%7B%222%22%3A%22((state_name%20LIKE%20'%25Dako%25')%20AND%20(state_name%20%3D%20'Washington'))%22%7D&SIZE=157%2C179&BBOX=-173.25855613587362%2C-69.66322093921872%2C-16.25855613587362%2C109.33677906078128&BBOXSR=4326&IMAGESR=4326&DPI=90"`;
